### PR TITLE
TcpStats.GetStats() now handles error when host cannot disclose information

### PR DESCRIPTION
**Pull request by RPM1984**
Hi! 👋 Long time consumer, first time contributor 😊 

I had a quick look at what uses these stats, and it appears the code can handle not having any stats, so I though this was the most reasonable approach.

But happy to update with whatever you guys prefer, equally happy to close this PR and leave to the experts if I'm way off! But, thought i'd try and give back if I can.

Fixes https://github.com/elastic/elasticsearch-net/issues/5176

Thanks!

### CODE REVIEW

**Comment by RPM1984** [comment]
Sorry, forgot to sign CLA. Done now, but not sure how I can "refresh" this PR with that?

File: src/Elastic.Transport/Diagnostics/TcpStats.cs
Comment (russcam) [inline src/Elastic.Transport/Diagnostics/TcpStats.cs:37]: since `GetActiveTcpConnections()` is a public method, I think the `try/catch` should be inside `GetActiveTcpConnections()`.

I think it ought to return `null` as you propose here, rather than `Array.Empty<TcpConnectionInformation>`, to imply that the information could not be retrieved rather than there were no connections

**Review by russcam — CHANGES_REQUESTED** [review]

File: src/Elastic.Transport/Diagnostics/TcpStats.cs
Comment (RPM1984) [reply src/Elastic.Transport/Diagnostics/TcpStats.cs:37]: Agreed! Will update now..

**Review by russcam — CHANGES_REQUESTED** [review]
This is almost there, I think. I left a couple of comments for consideration,

One other addition that would be good to make is to add a NOTE admonition to the DebugInformation documentation e.g.

https://github.com/elastic/elasticsearch-net/blob/110f68665b18c4952222f9c2baaf0732da65d55e/tests/Tests/ClientConcepts/Troubleshooting/DebugInformation.doc.cs#L223-L234

change to

```asciidoc
/**
* [NOTE]
* --
* Collecting TCP statistics may not be accessible in all environments, for example, Azure App Services.
* When this is the case, `TcpStats.GetActiveTcpConnections()` returns `null`.
* --
*
* ==== ThreadPool statistics
*
* It can often be useful to see the statistics for thread pool threads, particularly when
* trying to diagnose issues with the client. The client can collect statistics for both
* worker threads and asynchronous I/O threads, and expose these on the response and
* in debug information.
*
* Similar to collecting TCP statistics, ThreadPool statistics can be collected for all requests
* by configuring `EnableThreadPoolStats` on `ConnectionSettings`
*/
```

and run 

```
build.bat documentation 
```

to generate the asciidoc files in `./docs` from C# source

File: src/Elastic.Transport/Diagnostics/TcpStats.cs
Comment (RPM1984) [reply src/Elastic.Transport/Diagnostics/TcpStats.cs:44]: @russcam `EmptyReadOnly<TKey,TValue>.Dictionary` returns `IReadOnlyDictionary<TKey,TValue>`, but the signature on `GetStates` is `ReadOnlyDictionary`.

image

I can change to `IReadOnlyDictionary`, but.. that breaks a few things..

image

Happy to change, but wanted to get your OK first..

Comment (RPM1984) [reply src/Elastic.Transport/Diagnostics/TcpStats.cs:44]: can also do:

```
return new Dictionary<TcpState, int>(new Dictionary<TcpState, int>(0));
```

Which is a smaller change, but then there is some inconsistency, since we're not using that helper method..

Comment (russcam) [reply src/Elastic.Transport/Diagnostics/TcpStats.cs:44]: Ah, that's unfortunate. I think in next major, the method should be updated to return `IReadOnlyDictionary<TKey,TValue>`. Doing it in 7.x would be a breaking change so I think introducing a 

```
private static readonly ReadOnlyDictionary<TcpState, int> Empty = 
    new ReadOnlyDictionary<TcpState, int>(new Dictionary<TcpState, int>());
```

field and returning that would be OK.

Comment (RPM1984) [reply src/Elastic.Transport/Diagnostics/TcpStats.cs:44]: Agreed! Updating..

File: src/Elastic.Transport/Diagnostics/TcpStats.cs
Comment (RPM1984) [reply src/Elastic.Transport/Diagnostics/TcpStats.cs:26]: Fixed

File: src/Elastic.Transport/Diagnostics/TcpStats.cs
Comment (RPM1984) [reply src/Elastic.Transport/Diagnostics/TcpStats.cs:45]: Fixed.

Out of curiosity, how will you track updating the method signature in the next major? Should I create a _new_ issue to get that updated, tagged for the 8.x release, or is this something you guys will handle?

**Comment by RPM1984** [comment]
Not sure how to fix this check?

image

**Comment by russcam** [comment]
> Not sure how to fix this check?

Not sure if you can - don't worry about it

File: src/Elastic.Transport/Diagnostics/TcpStats.cs
Comment (stevejgordon) [reply src/Elastic.Transport/Diagnostics/TcpStats.cs:45]: I can open an issue to track that. We can take a PR for master without backporting to 7.x/7.10.

**Comment by stevejgordon** [comment]
This LGTM, thanks @RPM1984. @russcam Since you've been kindly reviewing, I'll let you have one final glance as well before merging.

File: src/Elastic.Transport/Diagnostics/TcpStats.cs
Comment (RPM1984) [reply src/Elastic.Transport/Diagnostics/TcpStats.cs:45]: @stevejgordon I'm happy to open a new PR for that if you like? Might need some guidance on the branching etc though, as not sure how a PR to master now will _not_ roll into the next 7.x release?

**Comment by RPM1984** [comment]
I had a look at the test failures, but.. they seem unrelated? Not sure if I missed a step, in the commit/PR process?

**Comment by russcam** [comment]
@RPM1984 the integration test failures do look unrelated - I think these should be addressed separately, so I've opened https://github.com/elastic/elasticsearch-net/issues/5185 to track

**Review by russcam — APPROVED** [review]
LGTM, thanks for your effort on this, @RPM1984 👍

**Comment by RPM1984** [comment]
> LGTM, thanks for your effort on this, @RPM1984 👍

My pleasure! Glad I could finally give back to this awesome lib. Keep up the awesome work @russcam @stevejgordon ❤️

**Comment by stevejgordon** [comment]
Thanks again @RPM1984. Really appreciate you contributing here.

**Pull request merged**
